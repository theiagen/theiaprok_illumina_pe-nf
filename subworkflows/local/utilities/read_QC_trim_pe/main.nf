include { FASTQ_SCAN_PE as FASTQ_SCAN_PE_RAW }         from '../../../../modules/local/quality_control/basic_statistics/fastq_scan/main'
include { FASTQ_SCAN_PE as FASTQ_SCAN_PE_CLEAN }         from '../../../../modules/local/quality_control/basic_statistics/fastq_scan/main'
include { FASTQC as FASTQC_RAW }              from '../../../../modules/local/quality_control/basic_statistics/fastqc/main'
include { FASTQC as FASTQC_CLEAN }              from '../../../../modules/local/quality_control/basic_statistics/fastqc/main'
include { READLENGTH }          from '../../../../modules/local/quality_control/basic_statistics/readlength/main'
include { BBDUK }               from '../../../../modules/local/quality_control/read_filtering/bbduk/main'
include { FASTP_PE }            from '../../../../modules/local/quality_control/read_filtering/fastp_pe/main'
include { NCBI_SCRUB_PE }       from '../../../../modules/local/quality_control/read_filtering/ncbi_scrub/main'
include { TRIMMOMATIC_PE }      from '../../../../modules/local/quality_control/read_filtering/trimmomatic/main'
include { KRAKEN2 }  from '../../../../modules/local/taxon_id/contamination/kraken2/main'
include { MIDAS }               from '../../../../modules/local/taxon_id/contamination/midas/run/main'

workflow READ_QC_TRIM_PE {
    // Keep path selected for theiaprok for first pass
    take:
    reads                    // channel: [ val(meta), path(read1), path(read2) ]
    trim_min_length          // val: minimum read length after trimming
    trim_quality_min_score   // val: minimum quality score for trimming
    trim_window_size         // val: window size for quality trimming
    call_midas               // val: boolean to call midas
    midas_db                 // path: midas database
    call_kraken              // val: boolean to call kraken
    kraken_db                // path: kraken database
    adapters                 // path: adapter sequences
    phix                     // path: phix sequences
    workflow_series          // val: workflow series (theiacov, theiaprok, etc.)
    read_processing          // val: read processing method (trimmomatic, fastp)
    read_qc                  // val: read qc method (fastq_scan, fastqc)
    trimmomatic_args         // val: trimmomatic arguments
    fastp_args               // val: fastp arguments

    main:
    ch_versions = Channel.empty()
    
    // Not sure if this is the best way to handle these channels, but it works for now
    // Just trying to map the outputs to the expected channels like we do in the wdl version
    ch_trimmed_reads = Channel.empty()
    ch_fastq_scan_raw_r1 = Channel.empty()
    ch_fastq_scan_raw_r2 = Channel.empty()
    ch_fastq_scan_raw_pairs = Channel.empty()
    ch_fastq_scan_raw_json1 = Channel.empty()
    ch_fastq_scan_raw_json2 = Channel.empty()
    ch_fastq_scan_clean_r1 = Channel.empty()
    ch_fastq_scan_clean_r2 = Channel.empty()
    ch_fastq_scan_clean_pairs = Channel.empty()
    ch_fastq_scan_clean_json1 = Channel.empty()
    ch_fastq_scan_clean_json2 = Channel.empty()
    ch_fastqc_raw_read1_html = Channel.empty()
    ch_fastqc_raw_read2_html = Channel.empty()
    ch_fastqc_clean_html1 = Channel.empty()
    ch_fastqc_clean_html2 = Channel.empty()
    ch_fastqc_clean_zip1 = Channel.empty()
    ch_fastqc_clean_zip2 = Channel.empty()
    ch_trimmomatic_stats = Channel.empty()
    ch_fastp_pe_html_report = Channel.empty()
    ch_midas_species_profile = Channel.empty()
    ch_kraken2_report = Channel.empty()
    ch_value_results = Channel.empty()

    // Read processing - trimmomatic or fastp
    if (read_processing == "trimmomatic") {
        TRIMMOMATIC_PE(
            reads,
            trim_min_length,
            trim_window_size,
            trim_quality_min_score,
            "", // base_crop - empty for no cropping
            trimmomatic_args ?: "-phred33"
        )
        ch_trimmed_reads = TRIMMOMATIC_PE.out.trimmed_reads
        ch_trimmomatic_stats = TRIMMOMATIC_PE.out.trimmomatic_stats
        ch_versions = ch_versions.mix(TRIMMOMATIC_PE.out.versions)
    }

    if (read_processing == "fastp") {
        FASTP_PE(
            reads,
            trim_window_size,
            trim_quality_min_score,
            trim_min_length,
            fastp_args ?: "--detect_adapter_for_pe -g -5 20 -3 20"
        )
        ch_trimmed_reads = FASTP_PE.out.trimmed_reads
        ch_fastp_pe_html_report = FASTP_PE.out.fastp_stats
        ch_versions = ch_versions.mix(FASTP_PE.out.versions)
    }

    // BBDuk cleaning
    BBDUK(
        ch_trimmed_reads,
        adapters ?: [],
        phix ?: []
    )
    ch_versions = ch_versions.mix(BBDUK.out.versions)

    ch_clean_reads = BBDUK.out.cleaned_reads

    // Quality control - fastqc or fastq_scan
    if (read_qc == "fastqc") {
        // FastQC on raw reads
        FASTQC_RAW (reads, "raw")
        ch_value_results = ch_value_results.mix(FASTQC_RAW.out.fastqc_value_results)
        ch_versions = ch_versions.mix(FASTQC_RAW.out.versions)
        ch_fastqc_raw_read1_html = FASTQC_RAW.out.read1_fastqc_html
        ch_fastqc_raw_read2_html = FASTQC_RAW.out.read2_fastqc_html

        // FastQC on clean reads
        FASTQC_CLEAN (ch_clean_reads, "clean")
        ch_value_results = ch_value_results.mix(FASTQC_CLEAN.out.fastqc_value_results)
        ch_fastqc_clean_html1 = FASTQC_CLEAN.out.read1_fastqc_html
        ch_fastqc_clean_html2 = FASTQC_CLEAN.out.read2_fastqc_html
        ch_fastqc_clean_zip1 = FASTQC_CLEAN.out.read1_fastqc_zip
        ch_fastqc_clean_zip2 = FASTQC_CLEAN.out.read2_fastqc_zip
        ch_versions = ch_versions.mix(FASTQC_CLEAN.out.versions)

    }

    if (read_qc == "fastq_scan") {
        // FastQ-scan on raw reads
        FASTQ_SCAN_PE_RAW (reads, "raw")
        ch_value_results = ch_value_results.mix(FASTQ_SCAN_PE_RAW.out.fastq_scan_value_results)
        ch_fastq_scan_raw_json1 = FASTQ_SCAN_PE_RAW.out.read1_fastq_scan_json
        ch_fastq_scan_raw_json2 = FASTQ_SCAN_PE_RAW.out.read2_fastq_scan_json
        ch_versions = ch_versions.mix(FASTQ_SCAN_PE_RAW.out.versions)

        // FastQ-scan on clean reads
        FASTQ_SCAN_PE_CLEAN (ch_clean_reads, "clean")
        ch_value_results = ch_value_results.mix(FASTQ_SCAN_PE_CLEAN.out.fastq_scan_value_results)
        ch_fastq_scan_clean_json1 = FASTQ_SCAN_PE_CLEAN.out.read1_fastq_scan_json
        ch_fastq_scan_clean_json2 = FASTQ_SCAN_PE_CLEAN.out.read2_fastq_scan_json
        ch_versions = ch_versions.mix(FASTQ_SCAN_PE_CLEAN.out.versions)
    }

    // MIDAS for theiaprok
    if (workflow_series == "theiaprok" && call_midas) {
        MIDAS(
            reads,
            midas_db
        )
        ch_midas_species_profile = MIDAS.out.species_profile
        ch_midas_log_file = MIDAS.out.log_file
        ch_value_results = ch_value_results.mix(MIDAS.out.midas_value_results)
        ch_versions = ch_versions.mix(MIDAS.out.versions)
    }

    // Kraken2 standalone for theiaprok
    if (workflow_series == "theiaprok" && call_kraken && kraken_db) {
        KRAKEN2(
            reads,
            kraken_db,
            "", // kraken2_args
            "classified#.fastq", // classified_out
            "unclassified#.fastq" // unclassified_out
        )
        ch_value_results = ch_value_results.mix(KRAKEN2.out.kraken2_percent_human_file)
        ch_kraken2_report = KRAKEN2.out.kraken2_report
        ch_versions = ch_versions.mix(KRAKEN2.out.versions)
    }

    
    emit:
    // BBDuk outputs
    bbduk_cleaned_reads       = BBDUK.out.cleaned_reads
    bbduk_adapters_stats      = BBDUK.out.adapter_stats
    bbduk_phix_stats          = BBDUK.out.phix_stats
    // FastQ-scan outputs
    fastq_scan_raw_r1        = ch_fastq_scan_raw_r1
    fastq_scan_raw_r2        = ch_fastq_scan_raw_r2
    fastq_scan_raw_pairs     = ch_fastq_scan_raw_pairs
    fastq_scan_clean_r1      = ch_fastq_scan_clean_r1
    fastq_scan_clean_r2      = ch_fastq_scan_clean_r2
    fastq_scan_clean_pairs   = ch_fastq_scan_clean_pairs
    fastq_scan_raw_json1     = ch_fastq_scan_raw_json1
    fastq_scan_raw_json2     = ch_fastq_scan_raw_json2
    fastq_scan_clean_json1   = ch_fastq_scan_clean_json1
    fastq_scan_clean_json2   = ch_fastq_scan_clean_json2
    // FastQC outputs
    fastqc_raw_html1         = ch_fastqc_raw_read1_html
    fastqc_raw_html2         = ch_fastqc_raw_read2_html
    fastqc_clean_html1       = ch_fastqc_clean_html1
    fastqc_clean_html2       = ch_fastqc_clean_html2
    fastqc_clean_zip1        = ch_fastqc_clean_zip1
    fastqc_clean_zip2        = ch_fastqc_clean_zip2
    // Kraken2 outputs
    kraken2_report           = ch_kraken2_report
    // Trimming outputs
    trimmomatic_stats        = ch_trimmomatic_stats
    fastp_html_report        = ch_fastp_pe_html_report
    // MIDAS outputs
    midas_species_profile    = ch_midas_species_profile
    // Versions
    versions                 = ch_versions
}