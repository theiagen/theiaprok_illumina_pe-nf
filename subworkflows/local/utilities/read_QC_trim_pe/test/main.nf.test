nextflow_workflow {

    name "Test Subworkflow READ_QC_TRIM_PE"
    script "../main.nf"
    workflow "READ_QC_TRIM_PE"

    test("read_qc_trim_pe - trimmomatic with fastqc") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = 50                           // trim_min_length
                input[2] = 20                           // trim_quality_min_score
                input[3] = 4                            // trim_window_size
                input[4] = false                        // call_midas
                input[5] = []                           // midas_db
                input[6] = false                        // call_kraken
                input[7] = []                           // kraken_db
                input[8] = []                           // adapters
                input[9] = []                           // phix
                input[10] = "theiaprok"                  // workflow_series
                input[11] = "trimmomatic"               // read_processing
                input[12] = "fastqc"                    // read_qc
                input[13] = "-phred33"                  // trimmomatic_args
                input[14] = ""                          // fastp_args
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.bbduk_cleaned_reads,
                    workflow.out.bbduk_adapters_stats,
                    workflow.out.bbduk_phix_stats,
                    workflow.out.fastqc_raw_html1,
                    workflow.out.fastqc_raw_html2,
                    workflow.out.fastqc_clean_html1,
                    workflow.out.fastqc_clean_html2,
                    workflow.out.trimmomatic_stats,
                    workflow.out.value_results,
                    workflow.out.versions).match() }
            )
        }
    }

    test("read_qc_trim_pe - fastp with fastq_scan") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = 30                           // trim_min_length
                input[2] = 15                           // trim_quality_min_score
                input[3] = 5                            // trim_window_size
                input[4] = false                        // call_midas
                input[5] = []                           // midas_db
                input[6] = false                        // call_kraken
                input[7] = []                           // kraken_db
                input[8] = []                          // adapters
                input[9] = []                           // phix
                input[10] = "theiaprok"                  // workflow_series
                input[11] = "fastp"                     // read_processing
                input[12] = "fastq_scan"                // read_qc
                input[13] = ""                          // trimmomatic_args
                input[14] = "--detect_adapter_for_pe -g -5 20 -3 20" // fastp_args
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.bbduk_cleaned_reads,
                    workflow.out.bbduk_adapters_stats,
                    workflow.out.bbduk_phix_stats,
                    workflow.out.fastqc_raw_html1,
                    workflow.out.fastqc_raw_html2,
                    workflow.out.fastqc_clean_html1,
                    workflow.out.fastqc_clean_html2,
                    workflow.out.trimmomatic_stats,
                    workflow.out.value_results,
                    workflow.out.versions).match() }
            )
        }
    }

    test("read_qc_trim_pe - theiaprok with midas and kraken") {

        setup {
            run ("UNTAR") {
                script "../../../../../modules/nf-core/untar/main.nf"
                process {
                    """
                    input[0] = [
                        [id:'db'],
                        [file('https://github.com/snayfach/MIDAS/raw/refs/heads/master/test/genomes.tar.gz', checkIfExists:true)]
                    ]
                    """
                }
            }
            run ("MIDAS_BUILD") {
                script "../../../../../modules/local/taxon_id/contamination/midas/build/main.nf"
                process {
                    """
                    input[0] = UNTAR.out.untar
                        .combine(
                            Channel.of(
                                file('https://github.com/snayfach/MIDAS/raw/refs/heads/master/test/genomes.mapfile', checkIfExists:true)
                            )
                        )
                    """
                }
            }
        }
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = 50                           // trim_min_length
                input[2] = 20                           // trim_quality_min_score
                input[3] = 4                            // trim_window_size
                input[4] = true                         // call_midas
                input[5] = MIDAS_BUILD.out.database.map { meta, db -> db }   // midas_db
                input[6] = true                         // call_kraken
                input[7] = file("https://github.com/nf-core/test-datasets/raw/refs/heads/taxprofiler/data/database/kraken2/testdb-kraken2.tar.gz")      // kraken_db
                input[8] = []                              // adapters
                input[9] = []                               // phix
                input[10] = "theiaprok"                 // workflow_series
                input[11] = "trimmomatic"               // read_processing
                input[12] = "fastqc"                    // read_qc
                input[13] = "-phred33"                  // trimmomatic_args
                input[14] = ""                          // fastp_args
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.bbduk_cleaned_reads,
                    workflow.out.bbduk_adapters_stats,
                    workflow.out.bbduk_phix_stats,
                    workflow.out.fastqc_raw_html1,
                    workflow.out.fastqc_raw_html2,
                    workflow.out.fastqc_clean_html1,
                    workflow.out.fastqc_clean_html2,
                    workflow.out.trimmomatic_stats,
                    workflow.out.versions).match() }
            )
        }
    }
}