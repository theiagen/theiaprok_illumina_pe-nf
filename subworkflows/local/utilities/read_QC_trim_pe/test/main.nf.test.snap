{
    "read_qc_trim_pe - theiaprok with midas and kraken": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample_1.clean.fastq.gz:md5,ec07275611622b5c4e554e07d3b4befc",
                        "test_sample_2.clean.fastq.gz:md5,e1617eb5740a628ed5af1d9bf185026f"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.adapters.stats.txt:md5,3e96d32762360eaf93bb0387c920d562"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.phix.stats.txt:md5,e05e344ed6b20d049571b28f74967099"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastqc.html:md5,f82da219c432ba1532c0d26fde37b6a5"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastqc.html:md5,16f62e3d51135e7ce6987ef2604a31a9"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastqc.html:md5,c5808ee09e1c6f9cd8b4e9a4ae3de31e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastqc.html:md5,dcf91c282af5c6c85cef0a5ee13f548b"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.trim.stats.txt:md5,fe6bf1d41433138650ebb20c6bcaa964"
                ]
            ],
            [
                "versions.yml:md5,6872bc8707d17cd0b07f35a96327f6b8",
                "versions.yml:md5,7964a5dc839d7476dc7d4a408011e6b5",
                "versions.yml:md5,8a91dc855526339f11edd06ae1803d98",
                "versions.yml:md5,8c6f2a5f43364b559b29d652487d672b",
                "versions.yml:md5,ae738d1bfe9b86b4d7e7c9925a5e2891",
                "versions.yml:md5,d0404a496925562eed55e0d39a517631"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T20:01:57.794483011"
    },
    "read_qc_trim_pe - fastp with fastq_scan": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample_1.clean.fastq.gz:md5,df2ed5fc0f91e00e4d3a28b8d22bbf88",
                        "test_sample_2.clean.fastq.gz:md5,e02b7c31208fea7d7cb2e83bdc613704"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.adapters.stats.txt:md5,dcc8b1db0bfc1730f7eddc19799b8b72"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.phix.stats.txt:md5,ed2c74827d798ccc2e53987d24dbf813"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,19f01fcc613b35ae4e45b60d433532b3",
                "versions.yml:md5,ae738d1bfe9b86b4d7e7c9925a5e2891",
                "versions.yml:md5,b7829b9f901ca8156c01d3bce78345c1",
                "versions.yml:md5,bde7af4cc4d94048fa6d36cc4aa7c112"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T19:46:01.00479296"
    },
    "read_qc_trim_pe - trimmomatic with fastqc": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample_1.clean.fastq.gz:md5,ec07275611622b5c4e554e07d3b4befc",
                        "test_sample_2.clean.fastq.gz:md5,e1617eb5740a628ed5af1d9bf185026f"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.adapters.stats.txt:md5,3e96d32762360eaf93bb0387c920d562"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.phix.stats.txt:md5,e05e344ed6b20d049571b28f74967099"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastqc.html:md5,f82da219c432ba1532c0d26fde37b6a5"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastqc.html:md5,16f62e3d51135e7ce6987ef2604a31a9"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastqc.html:md5,c5808ee09e1c6f9cd8b4e9a4ae3de31e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastqc.html:md5,dcf91c282af5c6c85cef0a5ee13f548b"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.trim.stats.txt:md5,fe6bf1d41433138650ebb20c6bcaa964"
                ]
            ],
            [
                "versions.yml:md5,6872bc8707d17cd0b07f35a96327f6b8",
                "versions.yml:md5,8a91dc855526339f11edd06ae1803d98",
                "versions.yml:md5,8c6f2a5f43364b559b29d652487d672b",
                "versions.yml:md5,ae738d1bfe9b86b4d7e7c9925a5e2891"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-23T19:45:09.061960651"
    }
}